tools/generate-sprintf-pow10-table.raku

```diff
@@ -0,0 +1,164 @@
+use v6;
+
+# The problem we have is that for floating point formatting, sprintf needs
+# accurate powers of 10, and nqp::num_n(1e1, $exp) is a call to the libc pow()
+# function, and *that* may not always be accurate, with the result that we
+# generate incorrect formatting. pow() is accurate (for all positive powers)
+# on the little endian Linux platforms I tested but on ppc64 and sparc64,
+# nqp::say(nqp::pow_n(10, 210)) returns 1.0000000000000001e+210
+# (which I think is an error of 1 bit in the last place of the mantissa)
+# However, MinGW seems to be much worse, with errors of 4 or 5 bits.
+#
+# The obvious approach would be to implement the algorithm from MoarVM's pow_i,
+# but in NQP in nums:
+#     while (exp) {
+#         if (exp & 1)
+#             result *= base;
+#         exp >>= 1;
+#         base *= base;
+#     }
+#
+# however, *this* doesn't work because floating point rounding errors from
+# *some* multiplications happen to propagate, causing higher powers to be
+# inaccurate. (and less accurate than the pow() function). Unlike pow, IEEE
+# basic arithmetic is defined to be bit-perfect correct, but that definition
+# includes rounding, and that can't be avoided.
+#
+# After some head scratching, it turns out that for all powers of 10, there
+# exist *some* smaller powers of 10 that can be multiplied together to calculate
+# them, without rounding errors, but not in any pattern that can be expressed
+# algorithmicly. However, we can generate a lookup table that stores which
+# smaller powers to multiply, and if we ensure that one is always a "small"
+# power which we can construct accurately as integers, then we can iterate our
+# way to any power of 10.
+#
+# This code generates and verifies that table. The table is pasted into
+# sprintf.nqp, where there is an NQP implementation of pow10()
+#
+# See https://github.com/MoarVM/MoarVM/pull/1385
+
+use nqp;
+
+# MoarVM decimal parsing is accurate. The platform's pow function might not be:
```

Comment:
Silly question: if MoarVM's decimal parsing is accurate and can generate the true values we're looking for, why not just hard code this table in the actual sprintf code?

Reply:
It works out at 2.4K of storage, most of which isn't needed by most programs, whereas I figured that about 440 bytes plus some other lookup code was about 25% of the size. To me, that felt a better trade off.